src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder.cs

```diff
@@ -402,21 +396,72 @@ internal static class DependentTypeFinder
             }
 
             var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
-            var typesInCompilation = GetAllAccessibleMetadataTypesInCompilation(compilation, cancellationToken);
-
             var result = new HashSet<INamedTypeSymbol>(SymbolEquivalenceComparer.Instance);
 
-            foreach (var type in typesInCompilation)
+            // Seed the current set of types we're searching for with the types we were given.
+            var currentTypes = metadataTypes;
+
+            while (currentTypes.Count > 0)
             {
-                if (metadataTypeMatches(metadataTypes, type))
+                var immediateDerivedTypes = new HashSet<INamedTypeSymbol>(SymbolEquivalenceComparer.Instance);
+
+                foreach (var reference in compilation.References.OfType<PortableExecutableReference>())
                 {
-                    result.Add(type);
+                    await FindImmediateMatchingMetadataTypesInMetadataReferenceAsync(
+                        currentTypes, project, metadataTypeMatches,
+                        compilation, reference, immediateDerivedTypes,
+                        cancellationToken).ConfigureAwait(false);
                 }
+
+                // Add what we found to the result set.
+                result.AddRange(immediateDerivedTypes);
+
+                // Now keep looping, using the set we found to spawn the next set of searches.
+                currentTypes = immediateDerivedTypes;
             }
 
             return result;
```

Comment:
It should not.  That would represent some sort of circular relationship *in metadata* (i.e. class A derives from B derives from A).  It's not something we have ever really concerned ourselves with (we have lots of other walking algorithms that would blow up if this could ever happen).